packages/shared/src/hooks/useUserMention.ts

```diff
@@ -0,0 +1,205 @@
+
+const getNode = (
+  el: HTMLElement,
+  query: string,
+): [ChildNode, { x: number; y: number }] => {
+  const index = Array.from(el.childNodes).findIndex((child) => {
+    const element = child.nodeValue ? child : child.childNodes[0];
+
+    if (isBreakLine(element)) {
+      return false;
+    }
+
+    const strings = element.nodeValue.split(' ');
+
+    return strings.some((string) => string === query);
+  });
+  const child = el.childNodes[index];
+  const node = child.nodeValue ? child : child.childNodes[0];
+  const start = node.nodeValue.indexOf(query);
+
+  return [node, { x: start, y: index }];
+};
+
+function setCaret(el: HTMLElement, replacement: string) {
+  const range = document.createRange();
+  const sel = window.getSelection();
+  const [node, { x }] = getNode(el, replacement);
+
+  range.setStart(node, x + replacement.length);
+  range.collapse(true);
+
+  sel.removeAllRanges();
+  sel.addRange(range);
+}
+
+function getCaretPost(el: Element): [number, number] {
+  const sel = window.getSelection();
+  const row = Array.from(el.childNodes).findIndex((child) => {
+    const element = child.nodeValue ? child : child.childNodes[0];
+
+    return sel.anchorNode === element;
+  });
+  return [sel.anchorOffset, row];
+}
+
+const getPossibleMentions = (el: HTMLElement) => {
+  const content = el.innerText.split(' ');
+  const mentions = content.filter((word) => word[0] === '@' && word.length > 1);
+
+  return mentions;
+};
+
+export function useUserMention({
+  postId,
+  commentRef,
+  onInput,
+}: UseUserMentionProps): UseUserMention {
+  const key = ['user-mention', postId];
+  const { user } = useContext(AuthContext);
+  const client = useQueryClient();
+  const [selected, setSelected] = useState(0);
+  const [offset, setOffset] = useState<[number, number]>([0, 0]);
+  const [mentions, setMentions] = useState<string[]>([]);
+  const [mentionQuery, setMentionQuery] = useState<string>();
+  const { data = { recommendedMentions: [] }, refetch } =
+    useQuery<RecommendedMentionsData>(
+      key,
+      () =>
+        request(`${apiUrl}/graphql`, RECOMMEND_MENTIONS_QUERY, {
+          postId,
+          query: mentionQuery,
+        }),
+      {
+        enabled: !!user,
+        refetchOnWindowFocus: false,
+        refetchOnMount: false,
+      },
+    );
+  const { recommendedMentions: mentionUsers } = data;
+  const fetchUsers = useDebounce(refetch, 300);
+  const filteredUsers = mentionUsers?.filter((recommendation) =>
+    mentions.every((mention) => `@${recommendation.username}` !== mention),
+  );
+
+  const onMention = (username: string) => {
+    const query = `@${mentionQuery}`;
+    const replacement = `@${username}`;
+    const element = commentRef?.current;
+    const content = element.innerHTML.replace(query, replacement);
+    // eslint-disable-next-line no-param-reassign
+    element.innerHTML = content;
+    setMentionQuery(undefined);
+    client.setQueryData(key, []);
```

Comment:
Not a 100% sure, but you might also want to invalidate the query at this point.
Perhaps we can test it to see if the query keeps refreshing for instance?

Reply:
I agree a better way would be invalidating the query - You are right, it won't request again due to invalidation since the `mentionQuery` is already `undefined`. Though we won't be able to actually create a test for it as it resides inside a tooltip which we actually can't test since then. This is why I had to test the specific component on its own.

Reply:
Unfortunately, even though the `mentionQuery` is already undefined. I have to keep setting it as an empty array to avoid the additional request and flick on the UI when invalidating.